numpy/_core/src/multiarray/multiarraymodule.c

```diff
@@ -1611,7 +1611,7 @@ _array_fromobject_generic(
         oldtype = PyArray_DESCR(oparr);
         if (PyArray_EquivTypes(oldtype, dtype)) {
             if (copy != NPY_COPY_ALWAYS && STRIDING_OK(oparr, order)) {
-                if (oldtype == dtype) {
+                if (oldtype == dtype || oldtype->kind == 'T') {
```

Comment:
Hmm, is this quite right? What if the user provides a dtype with a different `na_object`? I'd really expect `result.dtype == requested_dtype`... Or is does `EquivTypes` guarantee `na_object` is the same?

Reply:
Ah good point, you're right, we need to do something a little more complicated here, I'll revert this.